coap/adapter.go

```diff
@@ -1,61 +1,145 @@
+	Recieve(auth AuthProvider) MsgHandler
 }
 
-type CoAPAdapter struct {
-	obsMap map[string][]Observer
-	logger *zap.Logger
-	pub    mainflux.MessagePublisher
+const (
+	key       string = "key"
+	channel   string = "id"
+	protocol  string = "coap"
+	maxPktLen int    = 1500
+)
+
+var (
+	errBadRequest = errors.New("bad request")
+	logger        log.Logger
+)
+
+// AdapterService struct represents CoAP adapter service implementation.
+type AdapterService struct {
+	logger log.Logger
+	pubSub PubSub
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
+// New creates new CoAP adapter service struct.
+func New(lgr log.Logger, pubSub PubSub) Service {
+	logger = lgr
+	ca := &AdapterService{
 		logger: logger,
-		pub:    pub,
-		obsMap: make(map[string][]Observer),
+		pubSub: pubSub,
 	}
-
 	return ca
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
-}
+// Recieve method processes message and pushes it to NATS.
+func (ca *AdapterService) Recieve(auth AuthProvider) MsgHandler {
+	return func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message {
+		ca.logger.Log("message", fmt.Sprintf("Got message in Recieve: path=%q: %#v from %v", m.Path(), m, a))
+		var res *gocoap.Message
+
+		if m.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: m.MessageID,
+				Token:     m.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
 
-// BridgeHandler functions is a handler for messages recieved via NATS
-func (ca *CoAPAdapter) BridgeHandler(nm *broker.Msg) {
-	log.Printf("Received a message: %s\n", string(nm.Data))
+		if len(m.Payload) == 0 && m.IsConfirmable() {
+			res.Code = gocoap.BadRequest
+			return res
+		}
+
+		cid := mux.Var(m, channel)
+		publisher, err := authorize(m, res, cid, auth)
+		if err != nil {
+			ca.logger.Log("error", fmt.Sprintf("%s", err))
+			return res
+		}
 
-	// And write it into the database
-	m := mainflux.RawMessage{}
-	if len(nm.Data) > 0 {
-		if err := json.Unmarshal(nm.Data, &m); err != nil {
-			log.Println("Can not decode adapter msg")
-			return
+		n := mainflux.RawMessage{
+			Channel:   cid,
+			Publisher: publisher,
+			Protocol:  protocol,
+			Payload:   m.Payload,
 		}
+
+		if err := ca.pubSub.Publish(n); err != nil {
+			if m.IsConfirmable() {
+				res.Code = gocoap.InternalServerError
+			}
+			return res
+		}
+
+		if m.IsConfirmable() {
+			res.Code = gocoap.Changed
+		}
+		return res
 	}
+}
+
+// Observe method deals with CoAP observe messages.
+func (ca *AdapterService) Observe(auth AuthProvider) MsgHandler {
+	return func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message {
```

Comment:
While go enforces short names, it does not enforce meaningless names. Don't be cheap on characters :)